### PR TITLE
Add the process of Relationship V2 in EbeanLocalRelationshipWriterDAO

### PR DESCRIPTION
**Pull request by hongliu2024**
## Summary
This PR is to support `processLocalRelationshipUpdates` for relationship model v2.

## Details
Changes include:
1. GraphUtils.java
- Pass source asset urn in the `checkSameUrn` method for model v2.
2. GraphUtilsTest.java
- Add test cases of model v2 into all scenarios
3. ModelUtils.java
- Make `isRelationshipInV2` methods public so they can be used in EbeanLocalRelationshipWriterDAO logic
4. EbeanLocalRelationshipWriterDAO.java
- Add new parameter source asset urn to methods `addRelationships`, `addRelationshipGroup`, and `processRemovalOption`. The logic for model v1 remains the same, but the logic changes to take the passed parameter for model v2.
- Update documents
- Remove unused constant strings
5. EbeanLocalRelationshipWriterDAOTest.java
- Add model v2 test cases to RemoveFromSource test cases, since that's the one to be supported for V2.
6. ebean-local-relationship-dao-create-all.sql
- Add new relationship table based on the V2 relationship
7. ebean-local-relationship-create-all-with-non-dollar-virtual-column-names.sql
- Add new relationship table based on the V2 relationship

## Testing Done
1. ./gradlew build
```

BUILD SUCCESSFUL in 58s
245 actionable tasks: 37 executed, 208 up-to-date
```
2. Add V2 test cases

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java [hunk 1]
```diff
@@ -10,22 +10,35 @@
...
+   * @param relationships  list of relationships
+   * @param removalOption  removal option to specify which relationships to be removed
+   * @param sourceField    name of the source field
+   * @param destinationField name of the destination field
+   * @param urn  source urn to compare. Optional for V1. Needed for V2.
    */
   public static void checkSameUrn(@Nonnull final List<? extends RecordTemplate> relationships,
-      @Nonnull final BaseGraphWriterDAO.RemovalOption removalOption, final String sourceField, final String destinationField) {
+      @Nonnull final BaseGraphWriterDAO.RemovalOption removalOption, final String sourceField,
+      final String destinationField, Urn urn) {
```
Comment (ybz1013) [inline dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java:28]: nit: annotation for nullable params

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java [hunk 2]
```diff
@@ -10,22 +10,35 @@
...
 
     if (relationships.isEmpty()) {
       return;
     }
 
-    // ToDo: how to handle this for Relationship V2?
-    final Urn sourceUrn = getSourceUrnFromRelationship(relationships.get(0));
+    Urn sourceUrn = urn;
+    if (!ModelUtils.isRelationshipInV2(relationships.get(0).schema())) {
```
Comment (ybz1013) [inline dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java:35]: index check before .get(0) needed or no?

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java [hunk 3]
```diff
@@ -10,22 +10,35 @@
...
+      // get the sourceUrn from relationship, if relationship model in V1
+      sourceUrn = getSourceUrnFromRelationship(relationships.get(0));
+    }
+    if (sourceUrn == null) {
+      throw new IllegalArgumentException("Source urn is needed for Relationship V2");
```
Comment (ybz1013) [inline dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java:40]: The v1 v2 case is a little hard to follow. Could you add some comments or maybe consider this?
```
if (!ModelUtils.isRelationshipInV2(relationships.get(0).schema())) {
  // v1 case
} else {
  // v2 case
  if (urn == null) {
    throw new IllegalArgumentException("Source urn is needed for Relationship V2");
  }
  sourceUrn = urn;
}

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java [hunk 4]
```diff
@@ -85,21 +87,34 @@ public void clearRelationshipsByEntity(@Nonnull Urn urn,
...
+   * @param relationships the list of relationships to be persisted
+   * @param removalOption whether to remove existing relationship of the same type
+   * @param isTestMode whether to use test schema
+   * @param urn Urn of the entity to update relationships.
+   *            For Relationship V1: Optional, can be source or destination urn.
+   *            For Relationship V2: Required, is the source urn.
+   */
   public <RELATIONSHIP extends RecordTemplate> void addRelationships(@Nonnull List<RELATIONSHIP> relationships,
-      @Nonnull RemovalOption removalOption, boolean isTestMode) {
+      @Nonnull RemovalOption removalOption, boolean isTestMode, Urn urn) {
```
Comment (ybz1013) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:100]: nit: @Nullable Urn urn

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java [hunk 5]
```diff
@@ -133,12 +156,23 @@ private <RELATIONSHIP extends RecordTemplate> void addRelationshipGroup(@Nonnull
...
 
     long now = Instant.now().toEpochMilli();
 
     for (RELATIONSHIP relationship : relationshipGroup) {
-      Urn source = getSourceUrnFromRelationship(relationship);
+      // Relationship model V2 doesn't include source urn, it needs to be passed in.
+      // For relationship model V1, this given urn can be source urn or destination urn.
+      // For relationship model V2, this given urn can only be source urn.
+      Urn source = urn;
+      if (!ModelUtils.isRelationshipInV2(relationship.schema())) {
```
Comment (ybz1013) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:168]: duplicate code... Maybe a helper function `getSourceUrnFromEither(relationship, urn)`?

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java [hunk 6]
```diff
@@ -155,16 +189,28 @@ private <RELATIONSHIP extends RecordTemplate> void addRelationshipGroup(@Nonnull
...
+  private <RELATIONSHIP extends RecordTemplate> void processRemovalOption(@Nonnull String tableName,
+      @Nonnull RELATIONSHIP relationship, @Nonnull RemovalOption removalOption, Urn urn) {
 
     if (removalOption == RemovalOption.REMOVE_NONE) {
       return;
     }
 
     SqlUpdate deletionSQL = _server.createSqlUpdate(SQLStatementUtils.deleteLocalRelationshipSQL(tableName, removalOption));
-    Urn source = getSourceUrnFromRelationship(relationship);
+    Urn source = urn;
```
Comment (ybz1013) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:208]: ^

**Review by ybz1013 — COMMENTED** [review]
some comments. Thanks for adding this support!

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java (hunk 2 above)
Comment (hongliu2024) [reply dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java:35]: There is a null check at the begining of this method, so the index check won't be necessary.

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java (hunk 3 above)
Comment (hongliu2024) [reply dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java:40]: Thanks. Moved into a separate method and added comments.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java (hunk 4 above)
Comment (hongliu2024) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:100]: urn is nullable for V1

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java (hunk 5 above)
Comment (hongliu2024) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:168]: Good catch. Put that logic into a separate method, and removed the dups.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java (hunk 6 above)
Comment (hongliu2024) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:208]: same as the above. removed the dups. Thanks.

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java [hunk 7]
```diff
@@ -10,22 +10,28 @@
...
+   * @param relationships  list of relationships
+   * @param removalOption  removal option to specify which relationships to be removed
+   * @param sourceField    name of the source field
+   * @param destinationField name of the destination field
+   * @param urn  source urn to compare. Optional for V1. Needed for V2.
    */
   public static void checkSameUrn(@Nonnull final List<? extends RecordTemplate> relationships,
-      @Nonnull final BaseGraphWriterDAO.RemovalOption removalOption, final String sourceField, final String destinationField) {
+      @Nonnull final BaseGraphWriterDAO.RemovalOption removalOption, @Nonnull final String sourceField,
+      @Nonnull final String destinationField, Urn urn) {
```
Comment (jsdonn) [inline dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java:28]: nit: @Nonnull or @Nullable annotation for urn

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java [hunk 8]
```diff
@@ -38,9 +44,42 @@ public static void checkSameUrn(@Nonnull final List<? extends RecordTemplate> re
...
   }
 
+  /**
+   * Get the source asset's urn for a given relationship.
+   * @param relationship Relationship. The relationship can be in model V1 or V2.
+   * @param urn The source asset urn. Optional for V1. Must for V2. Exception will be thrown if urn is not provided for V2.
+   * @return The source asset urn.
+   */
+  public static <RELATIONSHIP extends RecordTemplate> Urn getSourceUrnBasedOnRelationshipVersion(
+      @Nonnull RELATIONSHIP relationship, Urn urn) {
```
Comment (jsdonn) [inline dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java:54]: nit: @Nonnull or @Nullable annotation for urn

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java [hunk 9]
```diff
@@ -38,9 +44,42 @@ public static void checkSameUrn(@Nonnull final List<? extends RecordTemplate> re
...
+
+  public static void checkSameUrn(@Nonnull final List<? extends RecordTemplate> relationships,
+      @Nonnull final BaseGraphWriterDAO.RemovalOption removalOption, final String sourceField, final String destinationField) {
+    checkSameUrn(relationships, removalOption, sourceField, destinationField, null);
+  }
+
   private static void checkSameUrn(@Nonnull List<? extends RecordTemplate> records, @Nonnull String field,
       @Nonnull Urn compare) {
     for (RecordTemplate relation : records) {
+      if (ModelUtils.isRelationshipInV2(relation.schema()) && field == SOURCE) {
```
Comment (jsdonn) [inline dao-api/src/main/java/com/linkedin/metadata/dao/utils/GraphUtils.java:78]: nit: string should use .equals comparison

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java [hunk 10]
```diff
@@ -85,21 +86,34 @@ public void clearRelationshipsByEntity(@Nonnull Urn urn,
...
     relationshipGroupMap.values().forEach(relationshipGroup
-        -> GraphUtils.checkSameUrn(relationshipGroup, removalOption, CommonColumnName.SOURCE, CommonColumnName.DESTINATION));
+        -> GraphUtils.checkSameUrn(relationshipGroup, removalOption, CommonColumnName.SOURCE, CommonColumnName.DESTINATION, urn));
 
     relationshipGroupMap.values().forEach(relationshipGroup -> {
-      addRelationshipGroup(relationshipGroup, removalOption, isTestMode);
+      addRelationshipGroup(relationshipGroup, removalOption, isTestMode, urn);
     });
   }
+  @Override
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:112]: nit: newline after closing curly brace

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java [hunk 11]
```diff
@@ -122,8 +136,16 @@ public <URN extends Urn> void removeEntities(@Nonnull List<URN> urns) {
...
+   * Add the given list of relationships to the local relationship tables.
+   * @param relationshipGroup the list of relationships to be persisted
+   * @param removalOption whether to remove existing relationship of the same type
+   * @param isTestMode  whether to use test schema
+   * @param urn the source urn to be used for the relationships. Optional for Relationship V1.
+   *            Needed for Relationship V2 because source is not included in the relationshipV2 metadata.
+   */
   private <RELATIONSHIP extends RecordTemplate> void addRelationshipGroup(@Nonnull final List<RELATIONSHIP> relationshipGroup,
-      @Nonnull RemovalOption removalOption, boolean isTestMode) {
+      @Nonnull RemovalOption removalOption, boolean isTestMode, Urn urn) {
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:148]: nit: @Nonnull or @Nullable annotation for urn

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java [hunk 12]
```diff
@@ -155,16 +180,23 @@ private <RELATIONSHIP extends RecordTemplate> void addRelationshipGroup(@Nonnull
...
+  /**
+   * Process the relationship removal in the DB tableName based on the removal option.
+   * @param tableName the table name of the relationship
+   * @param relationship the relationship to be removed
+   * @param removalOption the removal option
+   * @param urn the source urn to be used for the relationships. Optional for Relationship V1.
+   *            Needed for Relationship V2 because source is not included in the relationshipV2 metadata.
+   */
+  private <RELATIONSHIP extends RecordTemplate> void processRemovalOption(@Nonnull String tableName,
+      @Nonnull RELATIONSHIP relationship, @Nonnull RemovalOption removalOption, Urn urn) {
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:192]: nit: urn annotation

**Review by ybz1013 — APPROVED** [review]
LGTM

**Review by jsdonn — APPROVED** [review]
LGTM. thanks for the detailed PR description
